File: 0076 minimum-window-substring/solution.java
class Solution {
    public String minWindow(String s, String t) {
        int[] freq = new int[126];
        int required = 0;
        for (char i : t.toCharArray()) {
            required += freq[i] == 0 ? 1 : 0;
            freq[i]++;
        }
        int left = 0, right=0;
        int formed = 0, minLength = Integer.MAX_VALUE, start = 0;
        int[] window = new int[126];
        while (right < s.length()) {
            char c = s.charAt(right);
            window[c]++;
            if (freq[c] != 0 && (window[c] == freq[c]))
                formed++;
            while (left <=right && required == formed) {
                if (right - left + 1 < minLength) {
                    minLength = right - left + 1;
                    start = left;
                }
                char st = s.charAt(left);
                window[st]--;
                if (freq[st] > 0 && window[st] < freq[st]) {
                    formed--;
                }
                left++;
            }
            right++;
        }
        return minLength == Integer.MAX_VALUE ? "" : s.substring(start, start + minLength);
    }
}

File: 0076 minimum-window-substring/solution.cs
public class Solution {
    public string MinWindow(string s, string t) {

        int[] freq = new int[128];
        int req = 0;

        foreach(char c in t) {
            if(freq[c] == 0) req++;
            freq[c]++;
        }

        int l = 0, r = 0;
        int min = int.MaxValue;
        int start = 0;
        int formed = 0;

        int[] wd = new int[128];

        while(r < s.Length) {

            char c = s[r];
            wd[c]++;

            if(freq[c] != 0 && wd[c] == freq[c])
                formed++;

            while(l <= r && formed == req) {

                if(r - l + 1 < min) {
                    min = r - l + 1;
                    start = l;
                }

                char st = s[l];
                wd[st]--;

                if(freq[st] > 0 && wd[st] < freq[st])
                    formed--;

                l++;
            }

            r++;
        }

        return min == int.MaxValue ? "" : s.Substring(start, min);
    }
}
